WMF Framework/Event Platform Client/EPCNetworkManager.swift

```diff
@@ -0,0 +1,135 @@
+        
+        let task = session.dataTask(with: url) { (data, response, error) in
+            
+            let failureBlock = {
+                dispatchOnMainQueueAfterDelayInSeconds(attemptDelay) {
+                    self.httpDownload(url: url, retriesRemaining: retriesRemaining - 1, attemptDelay: attemptDelay, completion: completion)
+                }
+            }
+            
+            guard let httpResponse = response as? HTTPURLResponse,
+                let data = data else {
+                    DDLogWarn("EPCNetworkManager: server did not respond adequately, will try \(url.absoluteString) again")
+                failureBlock()
+                return
+            }
+            
+            guard httpResponse.statusCode == 200 else {
+                DDLogWarn("EPCNetworkManager: HTTP status of response was \(httpResponse.statusCode), will try \(url.absoluteString) again")
+                failureBlock()
+                return
+            }
+            
+            guard error == nil else {
+                failureBlock()
+                return
+            }
+            
+            completion(data)
+        }
+        task?.resume()
+    }
+    
+    func httpTryPost() {
+        var queuedEvent: (url: URL, body: NSDictionary)?
+        while !outputBufferIsEmpty() {
+            queuedEvent = removeOutputBufferAtIndex(0)
+            if let queuedEvent = queuedEvent {
+                self.submit(url: queuedEvent.url, payload: queuedEvent.body)
+            }
+        }
+    }
+
+    private func submit(url: URL, payload: Any? = nil) {
+        let request = session.request(with: url, method: .post, bodyParameters: payload, bodyEncoding: .json)
+        let task = session.dataTask(with: request, completionHandler: { (_, response, error) in
+            if error != nil {
+                DDLogError("EPCNetworkManager: An error occurred sending the request")
+            }
+        })
+        task?.resume()
+    }
+
+}
+
+private extension EPCNetworkManager {
+    /**
+     * Thread-safe asynchronous buffering of an event scheduled for POSTing
+     * - Parameter post: a tuple consisting of an `NSDictionary` `body` to be
+     *   POSTed to the `url`
+     */
+    func appendPostToOutputBuffer(_ post: (url: URL, body: NSDictionary)) {
+        queue.async {
+            self.outputBuffer.append(post)
+        }
+    }
+    /**
+     * Thread-safe synchronous check if any events have been scheduled
+     * - Returns: `true` if there are no scheduled evdents, `false` otherwise
+     */
+    func outputBufferIsEmpty() -> Bool {
+        queue.sync {
+            return self.outputBuffer.isEmpty
+        }
+    }
+    /**
+     * Thread-safe synchronous removal of scheduled event at index
+     * - Parameter index: The index from which to remove the scheduled event
+     * - Returns: a previously scheduled event
+     */
+    func removeOutputBufferAtIndex(_ index: Int) -> (url: URL, body: NSDictionary)? {
```

Comment:
You can also use `return self.outputBuffer.popFirst()` to clean this up. It returns an optional for in case the array is empty, so you can clean out the count checks.

Reply:
Yeah that sounds goood.

Reply:
```
Referencing instance method 'popFirst()' on 'Collection' requires the types '[(url: URL, body: NSDictionary)]' and 'Array<(url: URL, body: NSDictionary)>.SubSequence' (aka 'ArraySlice<(url: URL, body: NSDictionary)>') be equivalent
```

So I just changed it to:

``` swift
func outputBufferPopFirst() -> (url: URL, body: NSDictionary)? {
  queue.sync {
    if self.outputBuffer.isEmpty {
      return nil
    }
    return self.outputBuffer.remove(at: 0)
  }
}
```

Reply:
Lol... indeed how silly we all were to presume that a subset of an array be an array. How silly, when we know that Bread, once sliced, is Bread no longer, but is BreadSlice.